File: games/smite/smite1/god_builder.py
"""
GodBuilder for Smite 1.
Optimizes god builds for maximum DPS using OR-TOOLS.
"""

from typing import List, Optional, Tuple
from ortools.linear_solver import pywraplp

from games.smite.smite1.god import God
from games.smite.smite1.item import Item, Build, Starter


class GodBuilder:
    """
    Builder class for optimizing god item builds to maximize DPS.
    
    Supports:
    - Exactly 6 items in a build
    - Maximum 1 starter item (included in the 6 total)
    - DPS optimization using OR-TOOLS
    """
    
    def __init__(self, god: God, available_items: List[Item]):
        """
        Initialize the GodBuilder.
        
        Args:
            god: The god to build for
            available_items: List of available items to choose from
        """
        self.god = god
        self.available_items = available_items
    
    def calculate_dps(self, items: List[Item]) -> float:
        """
        Calculate DPS for a given set of items.
        
        Args:
            items: List of items equipped
            
        Returns:
            Damage per second
        """
        # Sum up stats from items
        power_physical = sum(item.stats.power_physical if item.stats else 0 for item in items)
        basic_attack_speed = sum(item.stats.basic_attack_speed if item.stats else 0 for item in items)
        
        # Add base god stats
        total_power = self.god.stats.power_physical + power_physical
        total_attack_speed = self.god.stats.basic_attack_speed + basic_attack_speed / 100
        
        # Calculate basic attack damage
        # Formula: base_damage + (power * scaling)
        basic_attack_damage = self.god.stats.basic_attack_damage + (self.god.basic_attack_scaling / 100) * total_power
        
        # Apply limits
        basic_attack_damage = min(basic_attack_damage, self.god.limits.basic_attack_damage_limit)
        total_attack_speed = min(total_attack_speed, self.god.limits.basic_attack_sec_limit)
        
        # DPS = damage per hit * hits per second
        dps = basic_attack_damage * total_attack_speed
        
        return dps
    
    def optimize_build(self) -> Optional[Tuple[Build, float]]:
        """
        Optimize the build for maximum DPS using OR-TOOLS.
        
        Since DPS = damage * attack_speed is non-linear, we use a weighted sum approach:
        We maximize a linear combination that approximates DPS maximization.
        
        Returns:
            Tuple of (Build, DPS) if successful, None otherwise
        """
        solver = pywraplp.Solver.CreateSolver('SCIP')
        if not solver:
            raise Exception("Solver not available")
        
        # === Binary decision variables for item inclusion ===
        item_vars = [solver.BoolVar(f'item_{i}') for i in range(len(self.available_items))]
        
        # === Variables for stats ===
        power_physical = solver.NumVar(0, solver.infinity(), 'power_physical')
        basic_attack_speed = solver.NumVar(0, solver.infinity(), 'basic_attack_speed')
        
        # === Constraints ===
        
        # 1. Exactly 6 items must be selected
        solver.Add(sum(item_vars) == 6)
        
        # 2. At most one Starter item
        starter_indices = [i for i, item in enumerate(self.available_items) if isinstance(item, Starter)]
        if starter_indices:
            solver.Add(sum(item_vars[i] for i in starter_indices) <= 1)
        
        # 3. Calculate total power from items
        solver.Add(
            power_physical == sum(
                item_vars[i] * (item.stats.power_physical if item.stats else 0) 
                for i, item in enumerate(self.available_items)
            )
        )
        
        # 4. Calculate total attack speed from items
        solver.Add(
            basic_attack_speed == sum(
                item_vars[i] * (item.stats.basic_attack_speed if item.stats else 0) 
                for i, item in enumerate(self.available_items)
            )
        )
        
        # === Objective: Maximize DPS approximation ===
        # Since we can't multiply variables (non-linear), we maximize a weighted sum
        # The weights are chosen to balance attack speed and power
        # DPS = (base_damage + power * scaling) * (base_AS + item_AS)
        # We approximate by maximizing: power * base_AS + item_AS * base_damage + power * item_AS
        # Simplified: maximize power + attack_speed with appropriate weights
        
        base_damage = self.god.stats.basic_attack_damage
        base_as = self.god.stats.basic_attack_speed
        scaling = self.god.basic_attack_scaling / 100
        
        # Weight power by (scaling * base_AS) and attack_speed by base_damage
        # This gives a linear approximation that favors balanced builds
        power_weight = scaling * base_as
        as_weight = base_damage
        
        solver.Maximize(power_weight * power_physical + as_weight * basic_attack_speed)
        
        # === Solve ===
        status = solver.Solve()
        
        if status == pywraplp.Solver.OPTIMAL:
            # Extract selected items
            selected_items = [
                self.available_items[i] for i in range(len(self.available_items)) 
                if item_vars[i].solution_value() > 0.5
            ]
            
            # Ensure we have exactly 6 items
            if len(selected_items) != 6:
                return None
            
            # Calculate actual DPS with selected items
            final_dps = self.calculate_dps(selected_items)
            
            # Create Build (fill positions based on whether they are starters or regular items)
            starter_items = [item for item in selected_items if isinstance(item, Starter)]
            regular_items = [item for item in selected_items if not isinstance(item, Starter)]
            
            # Pad with None if necessary
            while len(regular_items) < 5:
                regular_items.append(None)
            
            build = Build(
                item1=starter_items[0] if starter_items else regular_items.pop(0),
                item2=regular_items[0] if regular_items else None,
                item3=regular_items[1] if len(regular_items) > 1 else None,
                item4=regular_items[2] if len(regular_items) > 2 else None,
                item5=regular_items[3] if len(regular_items) > 3 else None,
                item6=regular_items[4] if len(regular_items) > 4 else None,
            )
            
            return build, final_dps
        
        return None

File: src/smite/god_builder.cpp
#include "smite/god_builder.h"
#include "ortools/linear_solver/linear_solver.h"
#include <algorithm>

namespace twoptimizer::smite {

GodBuilder::GodBuilder(const God& god, const std::vector<std::shared_ptr<Item>>& availableItems)
    : god_(god), available_items_(availableItems) {
}

double GodBuilder::calculateDps(const std::vector<std::shared_ptr<Item>>& items) const {
    // Calculate total stats with items
    double total_power = (god_.getPowerType() == PowerType::PHYSICAL) 
        ? god_.getStats().power_physical 
        : god_.getStats().power_magical;
    double total_attack_speed = god_.getStats().basic_attack_speed;
    
    for (const auto& item : items) {
        if (item) {
            if (god_.getPowerType() == PowerType::PHYSICAL) {
                total_power += item->getStats().power_physical;
            } else {
                total_power += item->getStats().power_magical;
            }
            total_attack_speed += item->getStats().basic_attack_speed;
        }
    }
    
    return total_power * (1.0 + total_attack_speed / 100.0);
}

std::optional<std::pair<Build, double>> GodBuilder::optimizeBuild() {
    if (available_items_.size() < 6) {
        return std::nullopt;
    }
    
    // Create OR-Tools solver
    auto solver = std::make_unique<operations_research::MPSolver>(
        "god_builder",
        operations_research::MPSolver::SCIP_MIXED_INTEGER_PROGRAMMING
    );
    
    // Create binary variables for each item
    std::vector<operations_research::MPVariable*> item_vars;
    for (size_t i = 0; i < available_items_.size(); ++i) {
        item_vars.push_back(solver->MakeBoolVar("item_" + std::to_string(i)));
    }
    
    // Constraint: exactly 6 items
    auto* constraint = solver->MakeRowConstraint(6.0, 6.0, "exactly_6_items");
    for (auto* var : item_vars) {
        constraint->SetCoefficient(var, 1.0);
    }
    
    // Constraint: at most 1 starter item
    auto* starter_constraint = solver->MakeRowConstraint(0.0, 1.0, "at_most_1_starter");
    for (size_t i = 0; i < available_items_.size(); ++i) {
        if (available_items_[i]->isStarter()) {
            starter_constraint->SetCoefficient(item_vars[i], 1.0);
        }
    }
    
    // Objective: maximize DPS (approximate with linear combination of stats)
    auto* objective = solver->MutableObjective();
    objective->SetMaximization();
    
    for (size_t i = 0; i < available_items_.size(); ++i) {
        const auto& stats = available_items_[i]->getStats();
        double power = (god_.getPowerType() == PowerType::PHYSICAL) 
            ? stats.power_physical 
            : stats.power_magical;
        
        // Simplified objective: power + attack_speed factor
        double coefficient = power + stats.basic_attack_speed * 0.5;
        objective->SetCoefficient(item_vars[i], coefficient);
    }
    
    // Solve
    auto result = solver->Solve();
    if (result != operations_research::MPSolver::OPTIMAL) {
        return std::nullopt;
    }
    
    // Extract solution
    std::vector<std::shared_ptr<Item>> selected_items;
    for (size_t i = 0; i < available_items_.size(); ++i) {
        if (item_vars[i]->solution_value() > 0.5) {
            selected_items.push_back(available_items_[i]);
        }
    }
    
    // Ensure we have exactly 6 items
    if (selected_items.size() != 6) {
        return std::nullopt;
    }
    
    // Create build
    Build build;
    build.item1 = selected_items.size() > 0 ? selected_items[0] : nullptr;
    build.item2 = selected_items.size() > 1 ? selected_items[1] : nullptr;
    build.item3 = selected_items.size() > 2 ? selected_items[2] : nullptr;
    build.item4 = selected_items.size() > 3 ? selected_items[3] : nullptr;
    build.item5 = selected_items.size() > 4 ? selected_items[4] : nullptr;
    build.item6 = selected_items.size() > 5 ? selected_items[5] : nullptr;
    
    double dps = calculateDps(selected_items);
    
    return std::make_pair(build, dps);
}

} // namespace twoptimizer::smite
